### PR TITLE
Add sequence number support for UCC

### PR DESCRIPTION
**Pull request by Fuzzkatt**
Add sequence number support for UCC, mostly following format of ProcressGroupNCCL.
Pass new test: `test_all_gather_object_subgroup`
Add skips for gather tests: `test_gather_object` and `test_gather_object_subgroup`

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

**Comment by ptrblck** [comment]
The lint issues are real, could you fix them, please?

File: torch/csrc/distributed/c10d/ProcessGroupUCC.cpp [hunk 1]
```diff
@@ -1569,6 +1571,12 @@ c10::intrusive_ptr<Work> ProcessGroupUCC::recv(
       "ucc:recv");
 }
 
+void ProcessGroupUCC::setSequenceNumberForGroup() {}
```
[the remark this answers is not in the document]
Comment (Fuzzkatt) [reply torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:1573]: The PrcoessGroupUcc `setSequenceNumber` and `getSequenceNumberForGroup` implementation here is based on the existing ProcessGroupNccl implementation: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L759-L763
Basically the `_seq` counter is initialized at 0, and it is only incremented during the `collective_post()` call, which is the equivalent to the ProcessGroupNccl handling during its equivalent `collective() `call: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1532

**Comment by Fuzzkatt** [comment]
Add barrier blocking fix from here: https://github.com/pytorch/pytorch/pull/86961

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -7694,12 +7696,14 @@ def _test_verify_model_across_rank(self, use_logger):
 
         @require_backend(DistTestCases.backend_feature["gpu"])
         @require_backends_available(DistTestCases.backend_feature["gpu"])
+        @sandcastle_skip_if(BACKEND == "ucc", "test failing locally with ucc")
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/distributed_test.py:7699]: Wondering what error this local failure would manifest to end users? 
Asking to see if it is safe to skip.

File: torch/csrc/distributed/c10d/ProcessGroupUCC.cpp [hunk 3]
```diff
@@ -170,6 +170,9 @@ void read_config() {
     torch_ucc_config.blocking_wait[(std::uint8_t)op] = true;
   }
 
+  // barrier is always blocking
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::BARRIER] = true;
+
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroupUCC.cpp:175]: Just FYI - this change is also in #86961

**Review by kwen2501 — COMMENTED** [review]
The code changes look good to me. 
Only have one question about the `@sandcastle_skip_if`

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (Fuzzkatt) [reply torch/testing/_internal/distributed/distributed_test.py]: Just updated the skip message. For both with and without logger versions, it times out after 5 minutes.

**Review by kwen2501 — APPROVED** [review]
Please create a GitHub issue for tracking fix to the skipped tests

**Pull request closed without merging**